### DISCUSSION TITLE
Is there any way to have dynamic fields?

### DISCUSSION THREAD

**Discussion by CleanCut** [tokio-rs/tracing#2253] [Q&A]
Is there any way to have dynamic fields? Something like:

```rust
let cloud_account_id = "cloud.account.id";
tracing::info!(cloud_account_id = 234, "created the cloud account id");
```

and have it output something with the `cloud.account.id` field and value `234`?

Even if it weren't technically "dynamic", but at least in some sort of constant, that'd be fine for me:

```rust
const CLOUD_ACCOUNT_ID: 'static str = "cloud.account.id";
tracing::info!(CLOUD_ACCOUNT_ID = 234, "created the cloud account id");
```

I tried making my own type and implementing `AsField`, which looked like it should have worked...

```rust
impl AsField for MyType {
    // ...
}

fn main() {
    let cloud_account_id = MyType::new("cloud.account.id");
    tracing::info!(cloud_account_id = 234, "created the cloud account id");
}
```

but `AsField` is a sealed trait 😭 , so I can't implement it outside `tracing` itself AFAICT.

---

If the answer is "No, there's no way to do that." Then my followup questions would be
- Is there any straightforward way to add such a feature?
  - Maybe something like this? `tracing::info!($cloud_account_id = 234, "created the cloud account id");`
- Or some workaround that I could do myself to kinda-sorta pretend that feature exists via my own macro(s) or something?

**Comment by bryangarza** [comment]
I believe that what you are describing (with the variable) is not possible, because the field names have to be static data (to minimize the number of allocations). Using a constant could potentially work if somewhere were to update the tracing macros to support it.

Q: Or some workaround that I could do myself to kinda-sorta pretend that feature exists via my own macro(s) or something?
A: Maybe if you have a macro that expands to the field name at compile-time `tracing::info!(cloud_account_id!() = 234, "created the cloud account id");`

**Reply by CleanCut** [reply]
I tried the workaround you suggested, but it looks like macros such as `info!` won't accept that format (either the `!` or the `()` or both, I'm guessing?)

```
error: format argument must be a string literal
  --> src/main.rs:31:20
   |
31 |     tracing::info!(cloud_account_id!() = 123, "created the cloud account");
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: you might be missing a string literal to format with
   |
31 |     tracing::info!("{} {}", cloud_account_id!() = 123, "created the cloud account");
   |                    ++++++++
```

**Reply by bryangarza** [reply]
Ah, okay. Yeah then seems it's not possible without modifying the macros to accept this sort of thing.

**Reply by CleanCut** [reply]
It looks like modifying the macros may work! I've got a basic POC working already in https://github.com/tokio-rs/tracing/pull/2254
